lib/utils/README.md

```diff
@@ -2,19 +2,375 @@
+    return lhs.first_name == rhs.first_name 
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+  }
 
+  std::string first_name;
+  std::string last_name;
+  int age;
+};
 ```
+If we take the previous example of adding an additional `is_male` field to `Person`, it can be easy to miss a location, leading to incorrectness. 
+For example, we could quite easily end up with
+```cpp
+struct Person {
+  Person() = delete;
+  Person(std::string const &first_name, 
+         std::string const &last_name, 
+         int age,
+         bool is_male)
+    : first_name(first_name),
+      last_name(last_name),
+      age(age),
+      is_male(is_male)
+    { }
+
+  friend bool operator==(Person const &lhs, Person const &rhs) {
+    return lhs.first_name == rhs.first_name 
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age
+      && lhs.is_male == rhs.is_male;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age;
+      // oops, forgot to update here. Have fun debugging :P
+  }
+
+  std::string first_name;
+  std::string last_name;
+  int age;
+  bool is_male;
+};
+```
+and for product types with more fields this grows increasingly tedious to write and maintain. 
+
+4. Hashing: hashing for product types is also relatively trivial, as long as each of the fields is hashable. But again, we have to do a bunch of extra work to specify this, and this work has to be done for each product type in the codebase.
+```cpp
+struct Person {
+  Person() = delete;
+  Person(std::string const &first_name, 
+         std::string const &last_name, 
+         int age,
+         bool is_male)
+    : first_name(first_name),
+      last_name(last_name),
+      age(age),
+      is_male(is_male)
+    { }
+
+  friend bool operator==(Person const &lhs, Person const &rhs) {
+    return lhs.first_name == rhs.first_name 
+      && lhs.last_name == rhs.last_name
+      && lhs.age == rhs.age
+      && lhs.is_male == rhs.is_male;
+  }
+
+  friend bool operator!=(Person const &lhs, Person const &rhs) {
+    return lhs.first_name != rhs.first_name 
+      || lhs.last_name != rhs.last_name
+      || lhs.age != rhs.age
+      || lhs.is_male != rhs.is_male;
+  }
+
+  std::string first_name;
+  std::string last_name;
+  int age;
+  bool is_male;
+};
+
+namespace std {
+
+template <>
+struct hash<::Person> {
+  size_t operator()(::Person const &p) const {
+    size_t result = 0;
+    hash_combine(result, p.first_name);
+    hash_combine(result, p.last_name);
+    hash_combine(result, p.age);
+    hash_combine(result, p.is_male);
+  }
+};
+
+}
+```
+and if we also want to support, say, `std::set<Person>`, we also have to add `operator<`
```

Comment:
I might be missing smt, but what does this operator example have to do with hashing?